File: cpp/Blind75/Stack/ValidParentheses.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        vector<char> stack;

        for (int i = 0; i < s.size(); ++i)
        {
            char c = s[i];
            if (c == '(' || c == '[' || c == '{')
            {
                stack.push_back(c);
            }
            else if (c == ')')
            {
                if (!stack.empty() && stack[stack.size() - 1] == '(')
                    stack.pop_back();
                else
                    return false;
            }
            else if (c == ']')
            {
                if (!stack.empty() && stack[stack.size() - 1] == '[')
                    stack.pop_back();
                else
                    return false;
            }
            else if (c == '}')
            {
                if (!stack.empty() && stack[stack.size() - 1] == '{')
                    stack.pop_back();
                else
                    return false;
            }
            else
            {
                return false;
            }
        }

        return stack.empty();
    }

    bool isValidOpt(string s) {
        stack<char> stack; 

        for (int i = 0; i < s.size(); ++i) {
            char c = s[i];
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.empty()) return false;
                char top = stack.top();
                if ((c == ')' && top == '(') || 
                    (c == ']' && top == '[') || 
                    (c == '}' && top == '{')
                ) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        return stack.empty();
    }
};

File: Python/Grind75/ValidParentheses.py
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if (len(stack) > 0) and (c == ')' and stack[-1] == '(' or c == '}' and stack[-1] == '{' or c == ']' and stack[-1] == '['):
                stack.pop()
            else:
                stack.append(c)

        return len(stack) == 0


class SolutionOptimized:
    def isValid(self, s: str) -> bool:
        stack = []
        pair = {')': '(', '}': '{', ']': '['}

        for c in s:
            if c in pair:
                if stack and stack[-1] == pair[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return len(stack) == 0


if __name__ == "__main__":
    solution = Solution()
    print(solution.isValid("()"))
    print(solution.isValid("()[]{}"))
    print(solution.isValid("(]"))
    print(solution.isValid("([)]"))
    print(solution.isValid("{[]}"))
